proposal/TommyLin/README.rst

```diff
@@ -0,0 +1,256 @@
+shot. Here are three photos to show the effect of aperture size.
+
++-------------------------------------+-------------------------------------+------------------------------------+
+| .. image:: wiki/Dof_blocks_f1_4.jpg | .. image:: wiki/Dof_blocks_f4_0.jpg | .. image:: wiki/Dof_blocks_f22.jpg |
+| Aperture=/1.4, DOF=0.8cm            | Aperture=f/4.0, DOF=2.2cm           | Aperture=f/22, DOF=12.4cm          |
++-------------------------------------+-------------------------------------+------------------------------------+
+*. Source: https://en.wikipedia.org/wiki/Depth_of_field
+
+Definition of depth of field (DOF)
+__________________________________
+
+.. image:: DOF_equation.png
+
++--------+----------------------+--------------------------------+
+| Symbol | Definition           | Example                        |
++========+======================+================================+
+| c      | circle of confusion  | Canon APS-C camera CoC=0.018mm |
++--------+----------------------+--------------------------------+
+| f      |  focal length        | EF100mm f/2.8L MACRO IS USM    |
++--------+----------------------+--------------------------------+
+| N      |  f-number            | f/2.8, f/4, f/10, f/16         |
++--------+----------------------+--------------------------------+
+| u      |  distance to subject | 5cm ~ 20cm for Camera          |
++--------+----------------------+--------------------------------+
+For example: A Canon APS-C camera with conditions in previous table and assume 
+N=10, u=10cm=100mm, this will result:
+
+**DOF = 0.36mm**
+
+The value of DOF=0.36mm is too small to get a clear image for object.
+Fortunately, photographer can take multiple **"digital"** photos and stack these
+photos together to make a focused photo in large area.
+
++-------------------------------------------------------+
+| .. image:: wiki/750px-Focus_stacking_Tachinid_fly.jpg |
++-------------------------------------------------------+
+*. Source: https://en.wikipedia.org/wiki/Focus_stacking
+
+
+Algorithm
+=========
+
+Edge detection
+______________
+
+Pixel values change fast (sharp) in the focused area. Pixel values chage gentlely
+(soft) in the defocused aread. The sharp aread consider as edge that pixel value
+varies fast. Follow steps aim to find the edges in a photo.
+
+1. Perform a **Gaussian blur** on image with following kernel:
+
+::
+
+   1  2  1
+   2  4  2
+   1  2  1
+
+2. Perform a **Laplacian Edge Detection kernel** on Gaussian Blurred image:
+
+::
+
+   -1 -1 -1
+   -1  8 -1
+   -1 -1 -1
+
+3. Take absolute value of Laplacian of Gaussian (LoG) result. This will quantify
+   the strength of edges with respect to the size and strength of kernel.
+4. Create a blank image, loop through each pixel and find the strongest edge in
+   the LoG(i.e. the highest value in the image stack) and take the RGB value for
+   that pixel from the corresponding image.
+
+Structural Similarity Index Measure (SSIM)
```

Comment:
How much computation do you expect the problems require?  If an estimate is not available at this stage, it is also good to come up with a strategy for pursuing the estimate.

Reply:
Basic information:

- Image size: 5184x3456 (RGB)
- Kernel size: 3x3
- Gaussian / Laplacian filtering => x2

For the focus stacking part there are roughly:
5184 x 3456 x 3 x [3 x 3] x 2 multiplication and addition per image
Then multiply with how many photos in the directory to stack.

For the SSIM part:
5184 x 3456 x 3 x (9 multiplication , 1 division, 6 addition)
Assume there are 4 photos in directory. To determinate the relationship of each other, above SSIM calculation needs to multiply with (4 x 3 / 2) = 6.

Reply:
Could you beef up the analysis and add it to the proposal?  It will be useful reference down the road.

Reply:
Okay! New section to evaluate computations will created.